client/fingerprint/vault.go

```diff
@@ -22,71 +23,90 @@ const (
+	}
 }
 
 func (f *VaultFingerprint) Fingerprint(req *FingerprintRequest, resp *FingerprintResponse) error {
-	config := req.Config
-
-	if config.VaultConfig == nil || !config.VaultConfig.IsEnabled() {
-		return nil
+	for _, cfg := range f.vaultConfigs(req) {
+		err := f.fingerprintImpl(cfg, resp)
+		if err != nil {
+			return err
+		}
 	}
+	return nil
+}
+
+// fingerprintImpl fingerprints for a single Vault cluster
+func (f *VaultFingerprint) fingerprintImpl(cfg *config.VaultConfig, resp *FingerprintResponse) error {
 
 	// Only create the client once to avoid creating too many connections to Vault
-	if f.client == nil {
-		vaultConfig, err := config.VaultConfig.ApiConfig()
+	client := f.clients[cfg.Name]
+	if client == nil {
+		vaultConfig, err := cfg.ApiConfig()
 		if err != nil {
-			return fmt.Errorf("Failed to initialize the Vault client config: %v", err)
+			return fmt.Errorf("Failed to initialize the Vault client config for %s: %v", cfg.Name, err)
 		}
-		f.client, err = vapi.NewClient(vaultConfig)
+		client, err = vapi.NewClient(vaultConfig)
 		if err != nil {
-			return fmt.Errorf("Failed to initialize Vault client: %s", err)
+			return fmt.Errorf("Failed to initialize Vault client for %s: %s", cfg.Name, err)
 		}
-		useragent.SetHeaders(f.client)
+		f.clients[cfg.Name] = client
+		useragent.SetHeaders(client)
 	}
 
 	// Connect to vault and parse its information
-	status, err := f.client.Sys().SealStatus()
+	status, err := client.Sys().SealStatus()
 	if err != nil {
-
 		// Print a message indicating that Vault is not available anymore
-		if f.lastState == vaultAvailable {
-			f.logger.Info("Vault is unavailable")
+		if lastState, ok := f.lastStates[cfg.Name]; !ok || lastState == vaultAvailable {
+			f.logger.Info("Vault is unavailable", "cluster", cfg.Name)
 		}
-		f.lastState = vaultUnavailable
+		f.lastStates[cfg.Name] = vaultUnavailable
 		return nil
 	}
 
-	resp.AddAttribute("vault.accessible", strconv.FormatBool(true))
-	// We strip the Vault prefix because < 0.6.2 the version looks like:
-	// status.Version = "Vault v0.6.1"
-	resp.AddAttribute("vault.version", strings.TrimPrefix(status.Version, "Vault "))
-	resp.AddAttribute("vault.cluster_id", status.ClusterID)
-	resp.AddAttribute("vault.cluster_name", status.ClusterName)
+	if cfg.Name == "default" {
+		resp.AddAttribute("vault.accessible", strconv.FormatBool(true))
+		resp.AddAttribute("vault.version", strings.TrimPrefix(status.Version, "Vault "))
+		resp.AddAttribute("vault.cluster_id", status.ClusterID)
+		resp.AddAttribute("vault.cluster_name", status.ClusterName)
+	} else {
+		resp.AddAttribute(fmt.Sprintf("vault.%s.accessible", cfg.Name), strconv.FormatBool(true))
+		resp.AddAttribute(fmt.Sprintf("vault.%s.version", cfg.Name), strings.TrimPrefix(status.Version, "Vault "))
+		resp.AddAttribute(fmt.Sprintf("vault.%s.cluster_id", cfg.Name), status.ClusterID)
+		resp.AddAttribute(fmt.Sprintf("vault.%s.cluster_name", cfg.Name), status.ClusterName)
+	}
```

Comment:
Is there value in also having `vault.default.*` attributes? I imagine that's what we want people to use moving forwards?

Reply:
I'm torn on this. That's what our RFC says, but in practice I suspect _most_ users will have 1 Vault (ex. all Nomad CE users will), which means we'd be sending a bunch of extra attributes no one cares about. And we can't _only_ send the `vault.default.*` b/c of backwards compatibility; a client updated before servers will appear to lose its Vault.

Reply:
Yeah, good point, the common case is likely going to be a single Vault cluster, so having a bunch of duplicate attributes sound unnecessary (and kind of confusing).